lib/compiler.js

```diff
@@ -155,52 +152,87 @@ Compiler.prototype.generate = function generate(code, options) {
-          'exp["' + exportName + '"] = ' + exportName + ';' +
-          'exp["' + exportName + '"].libs = _libs;'
-        ) + '}'),
 
       // Provide to YModules
-     'if (typeof modules === "object") {',
-       'modules.define("' + exportName + '",' +
-         JSON.stringify(deps.ym) + ',' +
-         'function(provide' + deps.ymVars + ') { ' +
+      'if (typeof modules === "object") {',
+
+      // re-define YM modules to require keys
+      deps.ymNames.length !== 0 ?
+        deps.ymNames.map(function (dep) {
+          var ymLib = deps.ym[dep];
+
+          return 'modules.define(' +
+            '"' + dep + '"' + ',' +
+            '["' + ymLib + '"],' +
+            'function (provide, ymlib) {' +
+              'provide(ymlib);' +
+            '});';
+        }).join(EOL) : '',
+
+      // wrap CommonJS deps
+      deps.commonJSNames.length !== 0 ?
+        deps.commonJSNames.map(function (dep) {
+          var commonPath = deps.commonJS[dep];
+
+          return 'modules.define(' +
+            '"' + dep + '"' + ',' +
+            'function (provide, prev) {' +
+              'provide(require("' + commonPath + '") || prev);' +
+            '});';
+        }).join(EOL) : '',
+
+      // wrap global deps
+      deps.globalNames.length !== 0 ?
+        deps.globalNames.map(function (dep) {
+          var globName = deps.global[dep];
+
+          return 'modules.define(' +
+            '"' + dep + '"' + ',' +
+            'function (provide, prev) {' +
+            'provide(glob["' + globName + '"] || prev);' +
+            '});';
+        }).join(EOL) : '',
+
+        'modules.define("' + exportName + '",' +
+          JSON.stringify(deps.ymLibs) + ',' +
+          'function(provide' + deps.ymVars + ') { ' +
             'var engine = buildBemXjst({' + deps.ymVarsString + '});' +
-            deps.ymLibs +
+            deps.ymEngineLibs +
             'provide(engine);' +
           '}' +
-       ');',
-     '}',
-
-      // Provide to global scope
-      deps.defineAsGlobal ? deps.globalNames.length !== 0 ? (
-        'var _libs = {};' +
-        deps.globalNames.map(function(dep) {
-          var globName = deps.global[dep];
-          return 'typeof glob["' + globName + '"] !== "undefined" && (' +
-            '_libs["' + dep + '"] = ' +
-            'glob["' + globName + '"]);';
-        }).join('') +
-        exportName + ' = buildBemXjst(_libs);' +
-        'exp["' + exportName + '"] = ' + exportName + ';' +
-        'exp["' + exportName + '"].libs = _libs;'
-      ) : (
-        exportName + '= buildBemXjst(this.global);' +
+        ');',
+      '} else {',
+
+        'var _libs = {};',
+
+        // Provide with CommonJS
+        deps.commonJSNames.length !== 0 ?
+            deps.commonJSNames.map(function(dep) {
+              var commonPath = deps.commonJS[dep];
+              return '_libs["' + dep + '"] = ' +
+                'glob && typeof glob["' + dep + '"] ' +
+                '!== "undefined" ? ' +
```

Comment:
looks like no need to do it (try to use global) here - it will get to us anyway on line 222